packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -3,59 +3,78 @@
- * @param visibleParticipants - Array containing currently rendered (visible) participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
- * @param MAX_TILES - Maximum number of tiles to calculate.
+ * Calculates the participants that should be renderd basis on the list of dominant
+ * speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call.
+ * {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers
+ * in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible)
+ * participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param maxTiles - Maximum number of tiles to calculate.
  * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
  */
 export const smartDominantSpeakerParticipants = (
   participants: VideoGalleryRemoteParticipant[] = [],
   dominantSpeakers: Array<string> = [],
   visibleParticipants: VideoGalleryRemoteParticipant[] = [],
-  MAX_TILES = 4
+  maxTiles = 4
 ): VideoGalleryRemoteParticipant[] => {
   if (!participants) return [];
 
-  // Only use the Max allowed dominant speakers.
-  const dominantSpeakerIds = dominantSpeakers.slice(0, MAX_TILES);
-
   // Don't apply any logic if total number of video streams is less than Max video streams.
-  if (participants.length <= MAX_TILES) return participants;
+  if (participants.length <= maxTiles) return participants;
 
-  // Initialize `visibleParticipants` if it is empty.
-  if (!visibleParticipants.length) {
-    visibleParticipants = participants.slice(0, MAX_TILES);
+  let currentParticipants = visibleParticipants.slice();
+  // Initialize `currentParticipants` if it is empty.
+  if (!currentParticipants.length) {
+    currentParticipants = participants.slice(0, maxTiles);
   }
 
-  const lastVisibleSpeakerIds = visibleParticipants.map((speaker) => speaker.userId);
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = Array.from(new Set(dominantSpeakers).values()).slice(0, maxTiles);
+  const lastVisibleSpeakerIds = currentParticipants.map((p) => p.userId);
   const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
 
   // Remove participants that are no longer dominant and replace them with new dominant speakers.
-  lastVisibleSpeakerIds.forEach((id, idx) => {
-    if (!dominantSpeakerIds.includes(id)) {
+  for (let index = 0; index < maxTiles; index++) {
+    const oldUserId = lastVisibleSpeakerIds[index];
+
+    if ((oldUserId && !dominantSpeakerIds.includes(oldUserId)) || oldUserId === undefined) {
```

Comment:
Isn't this the same as `oldUserId === undefined || !dominantSpeakerIds.includes(oldUserId)` ?
Is there a falsey value other than undefined that `oldUserId` could take?

Also nit: Even if this conditions stays the same, I'd check for `undefined` first

Reply:
Haha, yes it's the same. Guess I am failing my coding round.

Reply:
If `oldUserId` is empty string that would be the other falsey value. It _shouldn't_ be but the function could be misused as I don't see anything guarding that from happening. (not sure what should happen in this case just making sure its being considered)

Reply:
It can't/shouldn't be null ever. That would break the entire app cuz that means that one of our participants has an empty user ID
So I wouldn't worry about that here at all..